File: simlogiqueJSProto/js/core/Connector.js
class Connector {
	constructor() {
		this._connection = undefined;
		this._links = [];
	}
	
	_connect(link) { /* Friend of Link */
		this._links.push(link);
		if(this._connection != undefined)
			link.setConnection(this._connection);
	}
	
	_disconnect(link) { /* Friend of Link */
		var i = this._links.indexOf(link);
		this._links.splice(i, 1);
		if(this._links.length == 0)
			this.setConnection(undefined);
	}
	
	isConnected() {
		return this._links.length != 0;
	}
	
	setConnection(connection) {
		if(connection != this._connection) {
			this._connection = connection;
			for(var i = 0; i < this._links.length; i++)
				this._links[i].setConnection(connection);
		}
	}
	
	destructor() {
		while(this._links.length)
			this._links[0].destructor();
	}
}

File: SimLogiqueCpp/Connector.cpp
#include "Connector.h"

Connector::~Connector() {
    while(isConnected())
        delete(_links.back());
}

void Connector::setConnection(Connection* connection) {
    if(connection != _connection) {
        _connection = connection;
        for(auto it = _links.begin(); it != _links.end(); it++)
            (*it)->setConnection(connection);
    }
}

void Connector::connect(Link& link) {
    _links.push_back(&link);
    if(_connection != nullptr)
        link.setConnection(_connection);
}

void Connector::disconnect(Link& link) {
    auto it = std::find(_links.begin(), _links.end(), &link);
    _links.erase(it);
    if(!isConnected())
        setConnection(nullptr);
}
